lib/Dialect/AIEVec/Transforms/AIEVectorize.cpp

```diff
@@ -1824,6 +1828,289 @@ static void fuseFMAOpsForColumnTopology(func::FuncOp func, VectState *state) {
+// the acc of fma is a mul/fma operation which uses the same operands as fma.
+// the def of two operands are upd operations.
+// Transform -
+// %5 = aievec.mul %4, %0 {xoffsets = "0x03020100", xoffsets_hi = "0x07060504",
+// xsquare = "0x2110", xstart = "0", zoffsets = "0", zoffsets_hi = "0", zstart =
+// "0", zstep = "1"}
+//
+// %6 = aievec.mac %4, %0, %5 {xoffsets = "0x03020100",
+// xoffsets_hi = "0x07060504", xsquare = "0x2110", xstart = "2", zoffsets = "0",
+// zoffsets_hi = "0", zstart = "2", zstep = "1"}
+//
+// to-
+//
+// %7 = aievec.mul_conv %6, %1 {M = 16 : si32, N = 4 : si32}
+//
+// or transform the pattern like this-
+//
+// %9 = aievec.mac %8, %0, %6 {xoffsets = "0x03020100", xoffsets_hi =
+// "0x07060504", xsquare = "0x2110", xstart = "0", zoffsets = "0", zoffsets_hi =
+// "0", zstart = "4", zstep = "1"}
+//
+// %10 = aievec.mac %8, %0, %9 {xoffsets =
+// "0x03020100", xoffsets_hi = "0x07060504", xsquare = "0x2110", xstart = "2",
+// zoffsets = "0", zoffsets_hi = "0", zstart = "6", zstep = "1"}
+//
+// to-
+//
+// %9 =
+// aievec.fma_conv %8, %2, %7 {M = 16 : si32, N = 4 : si32}
+//
+// TODO: Currently, we only support mul_conv_16x4 and mac_conv_16x4 intrinsics
+// for AIE-ML architecture. More convolution intrinsics will be supported.
+static bool canFuseMulFMAOps(Operation *Op, VectState *state) {
+  // Check 1. This should be an aievec fma operation
+  assert(isa<aievec::FMAOp>(Op) && "operation must be an aievec fma op");
+  aievec::FMAOp curOp = cast<aievec::FMAOp>(Op);
+
+  // Check 2. Element type should be int16 or int8
+  VectorType vType = Op->getOperand(1).getType().cast<VectorType>();
+  Type stype = vType.getElementType();
+  IntegerType itype = stype.dyn_cast<IntegerType>();
+
+  if (!itype)
+    return false;
+
+  unsigned width = itype.getWidth();
+
+  if (!(width == 16 || width == 8))
+    return false;
+
+  // Check 3. Op must have three operands and one result
+  assert(Op->getNumOperands() == 3 && Op->getNumResults() == 1);
+
+  // Check 4. acc operand of the Op should be a mul op or fma op
+  Operation *mulOrFMAOp = Op->getOperand(2).getDefiningOp();
+
+  if (!isa<aievec::MulOp, aievec::FMAOp>(mulOrFMAOp))
+    return false;
+
+  // Check 5. mulOrFMAOp must have one use
+  if (!mulOrFMAOp->hasOneUse())
+    return false;
+
+  // Check 6. mulOrFMAOp and Op must have the same lhs and rhs
+  if (mulOrFMAOp->getOperand(0) != Op->getOperand(0) ||
+      mulOrFMAOp->getOperand(1) != Op->getOperand(1))
+    return false;
+
+  Value lhs = nullptr;
+  Value rhs = nullptr;
+  Value acc = nullptr;
+  bool isMulOp = false;
+
+  // If the acc operand is a mul op, we will try to generate mul_conv operation
+  // If the acc operand is a fma op, we will try to generate fma_conv operation
+  if (auto mulOp = dyn_cast<aievec::MulOp>(mulOrFMAOp)) {
+    isMulOp = true;
+    // Check 7. mulOp must have two operands and one result
+    assert(mulOp->getNumOperands() == 2 && mulOp->getNumResults() == 1);
```

Comment:
Is it even possible to hit this assert? Wouldn't the MulOp parser fail to parse the Op if it doesn't have two operands and one result?